packages/stokhos/test/UnitTest/Stokhos_SacadoMPVectorUnitTest.cpp

```diff
@@ -474,52 +516,73 @@ struct UnitTestSetup {
+  typedef UnitTestSetup<vec_type> UTS;                                                         \
+  __macro_for_tests__(__vec_type__,__scalar_type__)
 
-namespace StaticFixedVecTest {
-  typedef Kokkos::DefaultExecutionSpace execution_space;
-  typedef Stokhos::StaticFixedStorage<int,double,8,execution_space> storage_type;
-  typedef Sacado::MP::Vector<storage_type> vec_type;
-  typedef UnitTestSetup<vec_type> UTS;
-  VECTOR_UNIT_TESTS(StaticFixedVector)
+namespace DynamicVecTest
+{
+  TEST_DYNAMIC_STORAGE(DynamicStorage, DynamicVector, double, VECTOR_UNIT_TESTS_REAL_TYPE)
 }
 
-int main( int argc, char* argv[] ) {
-  Teuchos::GlobalMPISession mpiSession(&argc, &argv);
-
-  Kokkos::initialize();
-//  Kokkos::HostSpace::execution_space::initialize();
-//  if (!Kokkos::DefaultExecutionSpace::is_initialized())
-//    Kokkos::DefaultExecutionSpace::initialize();
+namespace DynamicStridedVecTest
+{
+  TEST_DYNAMIC_STORAGE(DynamicStridedStorage, DynamicStridedVector, double, VECTOR_UNIT_TESTS_REAL_TYPE)
+}
 
-  int res = Teuchos::UnitTestRepository::runUnitTestsFromMain(argc, argv);
+/**
+ * Common test structure for static storage.
+ */
+#define TEST_STATIC_STORAGE(__storage_type__,__vec_type__,__scalar_type__,__scalar_type_name__,__storage_size__,__macro_for_tests__)\
+    typedef ::Kokkos::DefaultExecutionSpace execution_space;                                                                            \
+    typedef ::Stokhos::__storage_type__<int,__scalar_type__,__storage_size__,execution_space> storage_type;                             \
+    typedef ::Sacado::MP::Vector<storage_type> vec_type;                                                                                \
+    typedef UnitTestSetup<vec_type> UTS;                                                                                              \
+    __macro_for_tests__(__vec_type__,__scalar_type_name__)
 
-  Kokkos::finalize();
-//  Kokkos::HostSpace::execution_space::finalize();
-//  if (Kokkos::DefaultExecutionSpace::is_initialized())
-//    Kokkos::DefaultExecutionSpace::finalize();
+namespace StaticVecTest
+{
+  TEST_STATIC_STORAGE(StaticStorage, StaticVector, double, double, 8, VECTOR_UNIT_TESTS_REAL_TYPE)
+}
 
-  return res;
+namespace StaticFixedVecTest
+{
+  namespace Double        {TEST_STATIC_STORAGE(StaticFixedStorage, StaticFixedVector,                   double ,                double, 8, VECTOR_UNIT_TESTS_REAL_TYPE   )}
+// Skip std::complex when using CUDA, because std::complex isn't supported in that case.
+#ifndef KOKKOS_ENABLE_CUDA
+  namespace Complex_std   {TEST_STATIC_STORAGE(StaticFixedStorage, StaticFixedVector,   std   ::complex<double>,    std_complex_double, 8, VECTOR_UNIT_TESTS_COMPLEX_TYPE)}
+#endif
+  namespace Complex_Kokkos{TEST_STATIC_STORAGE(StaticFixedStorage, StaticFixedVector, ::Kokkos::complex<double>, kokkos_complex_double, 8, VECTOR_UNIT_TESTS_COMPLEX_TYPE)}
```

Comment:
Probably necessary to put a guard around this line to carry out these tests only if Trilinos is being compiled with complex doubles.

Reply:
Also, it would probably be interesting to ask Eric wether we should test other types of storage as well.

Reply:
I think it's either/or `HAVE_TPETRA_INST_COMPLEX_DOUBLE` or `HAVE_TEUCHOS_COMPLEX`.  Probably interesting to check the cmake logic to see what gets set and whether one would be better than the other.

Reply:
Let's choose the teuchos one 😊

Reply:
OK

Reply:
Just make sure it is the same macro that defines the needed specializations.  You will probably need to use different macros for different kinds of specializations/instantiations, e.g., something for Stokhos itself (`HAVE_TEUCHOS_COMPLEX` should be fine as I think that is how Sacado does it), something else for any Kokkos specializations (should be the same on Kokkos uses), and `HAVE_TPETRA_INST_COMPLEX_DOUBLE` for Tpetra and downstream instantiations.

And you only need to due static, fixed storage.  I don't believe we even instantiate ensembles on the dynamic storage type anymore (it was just an intermediate step for building the PCE scalar type).

Reply:
Yes, although I think you would also disable `std::complex` for HIP.

Reply:
OK, let's remove `std::complex` for HIP as well :smile: